src/plugins/kdb/lmdb/kdb_lmdb.c

```diff
@@ -0,0 +1,1142 @@
+        key.mv_data = name;
+        key.mv_size = strlen(name);
+        klmdb_encode_princ_lockout(context, entry, lockout);
+        val.mv_data = lockout;
+        val.mv_size = sizeof(lockout);
+        err = mdb_txn_begin(dbc->lockout_env, NULL, 0, &txn);
+        if (!err && dbc->merge_nra) {
+            /* During an iprop load, do not change existing lockout entries. */
+            if (mdb_get(txn, dbc->lockout_db, &key, &dummy) == 0)
+                goto cleanup;
+        }
+        if (!err)
+            err = mdb_put(txn, dbc->lockout_db, &key, &val, 0);
+        if (!err) {
+            err = mdb_txn_commit(txn);
+            txn = NULL;
+        }
+        if (err) {
+            ret = klerr(context, err, _("LMDB lockout write failure"));
+            goto cleanup;
+        }
+    }
+
+cleanup:
+    mdb_txn_abort(txn);
+    krb5_free_unparsed_name(context, name);
+    return ret;
+}
+
+static krb5_error_code
+klmdb_delete_principal(krb5_context context, krb5_const_principal searchfor)
+{
+    krb5_error_code ret;
+    klmdb_context *dbc = context->dal_handle->db_context;
+    char *name;
+
+    if (dbc == NULL)
+        return KRB5_KDB_DBNOTINITED;
+
+    ret = krb5_unparse_name(context, searchfor, &name);
+    if (ret)
+        return ret;
+
+    ret = del(context, dbc->env, dbc->princ_db, name);
+    if (!ret && dbc->lockout_env != NULL)
+        (void)del(context, dbc->lockout_env, dbc->lockout_db, name);
+
+    krb5_free_unparsed_name(context, name);
+    return ret;
+}
+
+static krb5_error_code
+klmdb_iterate(krb5_context context, char *match_expr,
+              krb5_error_code (*func)(void *, krb5_db_entry *), void *arg,
+              krb5_flags iterflags)
+{
+    krb5_error_code ret = 0;
+    klmdb_context *dbc = context->dal_handle->db_context;
+    krb5_db_entry *entry;
+    MDB_txn *txn = NULL;
+    MDB_cursor *cursor = NULL;
+    MDB_val key, val;
+    MDB_cursor_op op = (iterflags & KRB5_DB_ITER_REV) ? MDB_PREV : MDB_NEXT;
+    int err;
+
+    if (dbc == NULL)
+        return KRB5_KDB_DBNOTINITED;
+
+    err = mdb_txn_begin(dbc->env, NULL, MDB_RDONLY, &txn);
+    if (err)
+        goto lmdb_error;
+    err = mdb_cursor_open(txn, dbc->princ_db, &cursor);
+    if (err)
+        goto lmdb_error;
+    for (;;) {
+        err = mdb_cursor_get(cursor, &key, &val, op);
+        if (err == MDB_NOTFOUND)
+            break;
+        if (err)
+            goto lmdb_error;
+        ret = klmdb_decode_princ(context, key.mv_data, key.mv_size,
+                                 val.mv_data, val.mv_size, &entry);
+        if (ret)
+            goto cleanup;
+        fetch_lockout(context, &key, entry);
+        ret = (*func)(arg, entry);
+        krb5_db_free_principal(context, entry);
+        if (ret)
+            goto cleanup;
+    }
+
+cleanup:
+    mdb_cursor_close(cursor);
+    mdb_txn_abort(txn);
+    return ret;
+lmdb_error:
+    ret = klerr(context, err, _("LMDB principal iteration failure"));
+    goto cleanup;
```

Comment:
I'm not really a fan of this use of `goto` for control flow... how would you feel about putting `lmdb_error:` above `cleanup:` for fall-through there and using an explicit `goto cleanup;`after the `for` loop?